### ISSUE TITLE
Speculation Rules API

### ISSUE THREAD

**Issue by BearCooder** [web-platform-tests/interop#754]
### Description

The Speculation Rules API is designed to improve performance for future navigations. It targets document URLs rather than specific resource files, and so makes sense for multi-page applications (MPAs) rather than single-page applications (SPAs).

The Speculation Rules API provides an alternative to the widely-available [<link rel="prefetch">](https://developer.mozilla.org/en-US/docs/Web/HTML/Attributes/rel/prefetch) feature. It provides many improvements over these technologies, along with a more expressive, configurable syntax for specifying which documents should be prefetched or prerendered.

### Specification

https://wicg.github.io/nav-speculation/speculation-rules.html

### Additional Signals

**Webkit:** https://github.com/WebKit/standards-positions/issues/54
**Mozilla:** https://github.com/mozilla/standards-positions/issues/620

**Tests:**
https://wpt.fyi/results/speculation-rules?label=master&label=experimental&aligned&q=speculation

Info about:
- https://developer.mozilla.org/en-US/docs/Web/API/Speculation_Rules_API
- https://developer.chrome.com/blog/speculation-rules-improvements

**Comment by domenic** [comment]
Oh, very cool to see this posted.

As one of the spec maintainers and Chromium engineers on this feature, I think it's important to emphasize that speculation rules is a surface API that triggers (as of now) two larger features: prefetching, or prerendering. Those are specified in https://wicg.github.io/nav-speculation/prefetch.html and https://wicg.github.io/nav-speculation/prerendering.html. So just implementing the speculation rules spec by itself doesn't help; you'd need to implement one or both of those.

If the parties involved don't want to accept everything into Interop 2025, I think accepting speculation rules + prefetching (omitting prerendering) would be a good compromise. As of October 1, speculation rules prefetch is used on [~12% of page loads](https://chromestatus.com/metrics/feature/timeline/popularity/4828), and this number is likely to go up a good chunk more as [Cloudflare's Speed Brain deployment continues](https://blog.cloudflare.com/introducing-speed-brain/). Speculation rules prerender, although it's more dramatic and certainly well-loved among web developers, [isn't seeing as-high adoption numbers](https://chromestatus.com/metrics/feature/timeline/popularity/3969). And working prefetch this year would be good groundwork if browsers wanted to tackle prerender the next year.

**Comment by ryantownsend** [comment]
@domenic I agree prefetch-only is a really good compromise on scope but one thought on this: would it be wise for them to treat prerender rules asif they were declared as prefetch?

In my view, prerender is a stronger signal that a URL/href will be the next page (I know we have eagerness for that too) so they _could_ depend on effectively downgrading prerender where it's not supported.

If browsers weren't to do this, then we'll end up having to conditionally serve Speculation Rules to different browsers (possibly needing UA sniffing?), which is far from ideal.

**Comment by domenic** [comment]
I agree that treating prerender as prefetch is reasonable. We've thought of doing that at times in Chromium (e.g. if limits are hit), but haven't as of yet.

**Comment by wpt-interop** [comment]
Thank you for proposing Speculation Rules API for inclusion in Interop 2025.

We wanted to let you know that this proposal was not selected to be part of Interop this year.

On behalf of the entire Interop team, thank you for submitting this proposal for consideration. We got many more proposals than we could include in this year's project, necessitating some difficult choices. Please note this should not be taken as a comment on the technology as a whole, or our willingness to consider it for Interop in the future. We appreciate the work you put into your proposal, and would welcome your participation in future rounds of Interop.

For an overview of our process, see [proposal selection](https://github.com/web-platform-tests/interop/blob/main/2025/selection-process.md). Thank you again for contributing to Interop 2025.

_Posted on behalf of the Interop team._
